Add smb backend

aneesh-n opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->
Adds SMB as a new storage backend in restic. Currently for backing up to SMB, it has to be mounted on the OS or rclone needs to be used. With this change it will be possible to specify SMB details and restic will directly make SMB connections to do the backups.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
Closes https://github.com/restic/restic/issues/4185

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].
-->

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

aneesh-n committed:
Add smb changes (#6)

* backend/smb: Add SMB backend and testcases

Add new SMB storage backend for restic.
Added test cases for testing SMB backend.

---------

Co-authored-by: Aneesh Nireshwalia <aneeshynot@gmail.com>
Co-authored-by: Srigovind Nayak <sgovind.dev@outlook.com>

Commit:
Update tests and run go fmt

Commit:
Add unrelease issue and cleanup configs

Removed extra environment variables

Commit:
Tidy go mod

Commit:
Add document and cleanup config

Add documentation for configuring smb repository.
Clean up configuration for smb. Renamed address to host.
Add option to configure user in smb repo url as well.
Options take highest precendence.

aneesh-n commented:
I have hardcoded smb credentials as it runs as a local test by installing samba on linux and creating an SMBShare for windows. MacOS smb test set up is pending.
Seems to be some issue with the tests, I will fix them and update.

Commit:
Fix test cases

Fix the check for macOS which was failing after commenting brew install.

Commit:
Fix lint issues

greatroar commented on internal/backend/smb/conpool.go:
```diff
@@ -0,0 +1,222 @@
+	return atomic.LoadInt32(&b.sessions)
+}
+
+// dial starts a client connection to the given SMB server. It is a
+// convenience function that connects to the given network address,
+// initiates the SMB handshake, and then sets up a Client.
```
There is no Client type in this code.

aneesh-n replied:
Will correct to - 'initiates the SMB handshake, and then returns a session for SMB communication.'

aneesh-n replied:
Regarding the copyright notice, does it make sense to include the following comment on top of conpool.go and smb.go -

// Parts of this code have been copied from Rclone (https://github.com/rclone)
// Copyright (C) 2012 by Nick Craig-Wood http://www.craig-wood.com/nick/

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

aneesh-n replied:
Updated comment. Added copyright notice.

greatroar commented:
Parts of this code have been copied from rclone. That means the [rclone license](https://github.com/rclone/rclone/blob/master/COPYING) should appear somewhere in it:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

greatroar commented on internal/backend/smb/smb.go:
```diff
@@ -0,0 +1,557 @@
+// Multiple programs or goroutines calling CreateTemp simultaneously will not choose the same file.
+// The caller can use the file's Name method to find the pathname of the file.
+// It is the caller's responsibility to remove the file when it is no longer needed.
+func (b *Backend) CreateTemp(cn *conn, dir, pattern string) (*smb2.File, error) {
+	if dir == "" {
+		dir = os.TempDir()
+	}
+
+	prefix, suffix, err := prefixAndSuffix(pattern)
+	if err != nil {
+		return nil, &fs.PathError{Op: "createtemp", Path: pattern, Err: err}
+	}
+	prefix = joinPath(dir, prefix)
+
+	try := 0
+	for {
+		name := prefix + nextRandom() + suffix
+		f, err := cn.smbShare.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0600)
+
+		if os.IsExist(err) {
+			if try++; try < 10000 {
+				continue
+			}
+			return nil, &fs.PathError{Op: "createtemp", Path: prefix + "*" + suffix, Err: fs.ErrExist}
+		}
+		return f, err
+	}
+}
```
Just do what the SFTP backend does:

```
// tempSuffix generates a random string suffix that should be sufficiently long
// to avoid accidental conflicts.
func tempSuffix() string {
        var nonce [16]byte
        _, err := rand.Read(nonce[:])
        if err != nil {
                panic(err)
        }
        return hex.EncodeToString(nonce[:])
}

tmpFilename := filename + "-restic-temp-" + tempSuffix()
```

aneesh-n replied:
Fixed

Commit:
Document env variables

Commit:
Fix review comments for temp file creation

Add rclone copyright notice for smb files.
Change temp file creation code to match sftp code.
Remove fastrand dependency.

Commit:
Move copyright notice below imports

Commit:
Add SecretString support for options

Commit:
Correct formatting

Commit:
Increase default max SMB connections to 5

greatroar commented on internal/backend/smb/conpool.go:
```diff
@@ -0,0 +1,243 @@
+
+// Return a SMB connection to the pool
+//
+// It nils the pointed to connection out so it can't be reused
+func (b *Backend) putConnection(pc **conn) {
```
Setting *pc back to nil is too easily defeated to be useful:

```
c, _ := b.getConnection(ctx, share)
c2 := c
b.putConnection(&c)
// c is nil, but c2 still refers to the conn.
```

I suggest just passing the pointer to this function. That saves some code and prevents the pointer from getting heap-allocated.

aneesh-n replied:
Got it. Thank you!

Commit:
Use connection pointer directly in putConnection

Setting *pc back to nil is too easily defeated to be useful.
This is more concise and prevents pointer from getting heap-allocated.

Commit:
Fix SMB test setup for Mac

Commit:
Fix SMB setup for linux

Corrected the variable name for smbuser

Commit:
Correct help for default smb Connections to 5

Commit:
Merge remote-tracking branch 'old-origin/master' into add-smb-backend

Commit:
Enable smb test for go 1.19

prajwaltr93 reviewed (approved):
Changes look good.

Commit:
Merge remote-tracking branch 'old-origin/master' into add-smb-backend

Commit:
Merge remote-tracking branch 'old-origin/master' into add-smb-backend

Commit:
fix linter warnings.
Correct method definition for unused param ctx

Commit:
Merge remote-tracking branch 'old-origin/master' into add-smb-backend

Commit:
Merge remote-tracking branch 'old-origin/master' into add-smb-backend

Commit:
fix lint - unused param prefix now used for smb

Commit:
Merge remote-tracking branch 'origin/master' into add-smb-backend

aneesh-n committed:
Merge remote-tracking branch 'restic-origin/master' into add-smb-backend

aneesh-n committed:
Merge branch 'master' into add-smb-backend

aneesh-n committed:
Fix lint issue in go.mod

aneesh-n committed:
Merge branch 'master' into add-smb-backend

aneesh-n committed:
Update go.mod

aneesh-n committed:
Update go.mod

aneesh-n committed:
Fix go.mod

aneesh-n committed:
Merge branch 'master' into add-smb-backend

aneesh-n committed:
Fix formatting in go mod

aneesh-n committed:
Merge remote-tracking branch 'origin/master' into add-smb-backend

aneesh-n committed:
Fix merge issues and update smb code similar to local

aneesh-n committed:
Fix go.mod and remove old go version tests

aneesh-n committed:
Refactor smb and local to use common helper

aneesh-n committed:
Fix lint and remove function

aneesh-n committed:
Clean up code and fix documentation

aneesh-n committed:
Fix lint and parse config logic

aneesh-n committed:
Update latest fixes from rclone SMB

Update documentation. Use cloudsoda fork of smb library for fixes related to size calculations and allowing to use hostnames.
